extend UserValue {
    optional int32 UserValue_int32 = 1;
    optional float UserValue_float = 2;
}

/////////////////////////////////////////

message UserValue {
    extensions 1 to 18999;
}

message UserError {
    extensions 1 to 18999;
}

message Get {
    required uint32 messageId = 1;
    required uint32 objectId = 2;
    required uint32 interfaceId = 3;
    required uint32 id = 4; // attribute
}

message Set {
    required uint32 messageId = 1;
    required uint32 objectId = 2;
    required uint32 interfaceId = 3;
    required uint32 id = 4; // attribute
    required UserValue value = 5;
}

message Subscribe {
    enum AttributeOrBroadcast {
        ATTRIBUTE = 1;
        BROADCAST = 2;
    }

    required uint32 messageId = 1;
    required uint32 objectId = 2;
    required uint32 interfaceId = 3;
    required uint32 id = 4; // attribute or broadcast
}

message Unsubscribe {
    enum AttributeOrBroadcast {
        ATTRIBUTE = 1;
        BROADCAST = 2;
    }

    required uint32 messageId = 1;
    required uint32 objectId = 2;
    required uint32 interfaceId = 3;
    required uint32 id = 4; // attribute or broadcast
}

message Fire {
    required uint32 messageId = 1;
    required uint32 objectId = 2;
    required uint32 interfaceId = 3;
    required uint32 id = 4; // method
    repeated UserValue in = 5; // arguments
}

message Broadcast {
    required uint32 objectId = 1;
    required uint32 interfaceId = 2;
    required uint32 id = 3; // attribute or broadcast
    repeated UserValue out = 4; // attribute value or broadcast output parameters
}

message Reply {
    enum RpcError {
        NO_SUCH_OBJECT = 1;
        NO_SUCH_INTERFACE = 2;
        BUFFER_OVERFLOW = 3;
        TYPE_SIGNATURE_MISMATCH = 4;
        FORBIDDEN = 5;
        OTHER = 6;
    }

    required uint32 inReplyToMessageId = 1;

    repeated UserValue out = 2; // output parameters
    optional UserError error = 3;
    optional RpcError rpcError = 4;
}

message RpcMessage {
    enum Type {
        GET = 1;
        SET = 2;
        SUBSCRIBE = 3;
        UNSUBSCRIBE = 4;
        FIRE = 5;
        BROADCAST = 6;
        REPLY = 7;
    }

    required Type type = 1;

    optional Get get = 16;
    optional Set set = 17;
    optional Subscribe subscribe = 18;
    optional Unsubscribe unsubscribe = 19;
    optional Fire fire = 20;
    optional Broadcast broadcast = 21;
    optional Reply reply = 22;
}
